### PR TITLE
[SPARK-45135][PYTHON][TESTS] Make `utils.eventually` a parameterized decorator

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?

- Make utils.eventually a parameterized decorator
- Retry `test_read_images` if it fails

### Why are the changes needed?
previously, we used `utils.eventually` to retry flaky tests, e.g. https://github.com/apache/spark/commit/745ed93fe451b3f9e8148b06356c28b889a4db5a

however, it needs to modify the test body, sometime the change maybe large

To minimize the changes, I'd like to ~~add a decorator for test retry~~ Make utils.eventually a parameterized decorator.

### Does this PR introduce _any_ user-facing change?
no, test-only

### How was this patch tested?
CI

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: python/pyspark/testing/utils.py [hunk 1]
```diff
@@ -115,6 +115,36 @@ def eventually(condition, timeout=30.0, catch_assertions=False):
         )
 
 
+def retry(maxTries=10, interval=1.0):
```
Comment (HyukjinKwon) [inline python/pyspark/testing/utils.py:118]: we have a similar utility `eventually` at `pyspark.testing.utils`.

Comment (zhengruifeng) [reply python/pyspark/testing/utils.py:118]: I see, I mentioned it in the PR description.
I think it will be more easy to use if we make it a decorator

Comment (HyukjinKwon) [reply python/pyspark/testing/utils.py:118]: I see. Can you make `eventually` to work as a decorator too?

Comment (zhengruifeng) [reply python/pyspark/testing/utils.py:118]: sound good, let me have a try

Comment (zhengruifeng) [reply python/pyspark/testing/utils.py:118]: it is a bit difference that `eventually` requires the function return `True` if succeed, so I will make it a different decorator

Comment (HyukjinKwon) [reply python/pyspark/testing/utils.py:118]: Can we reuse `eventually` here?

Comment (zhengruifeng) [reply python/pyspark/testing/utils.py:118]: hmm, maybe we can add an additional parameter to treat returned `None` as `True`

**Comment by zhengruifeng** [comment]
In the download log artifact, the added 3 UT:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="pyspark.tests.test_util.UtilTests-20230913111701" tests="6" file="pyspark/tests/test_util.py" time="7.334" timestamp="2023-09-13T11:17:09" failures="0" errors="0" skipped="0">
	<testcase classname="pyspark.tests.test_util.UtilTests" name="test_eventually_decorator" time="4.326" timestamp="2023-09-13T11:17:06" file="python/pyspark/tests/test_util.py" line="87">
		<system-out><![CDATA[
Attempt #1 failed!
False is not true

Attempt #2 failed!
False is not true

Attempt #3 failed!
False is not true

Attempt #4 failed!
False is not true

Attempt #5 failed!
False is not true

Attempt #6 failed!
False is not true

Attempt #7 failed!
False is not true

Attempt #8 failed!
False is not true

Attempt #9 failed!
False is not true

Attempt #10 failed!
False is not true

Attempt #11 failed!
False is not true

Attempt #12 failed!
False is not true

Attempt #13 failed!
False is not true
]]></system-out>
	</testcase>
	<testcase classname="pyspark.tests.test_util.UtilTests" name="test_eventually_function" time="0.610" timestamp="2023-09-13T11:17:06" file="python/pyspark/tests/test_util.py" line="93">
		<system-out><![CDATA[
Attempt #1 failed!
False is not true

Attempt #2 failed!
False is not true

Attempt #3 failed!
False is not true

Attempt #4 failed!
False is not true
]]></system-out>
	</testcase>
	<testcase classname="pyspark.tests.test_util.UtilTests" name="test_eventually_lambda" time="0.602" timestamp="2023-09-13T11:17:07" file="python/pyspark/tests/test_util.py" line="101">
		<system-out><![CDATA[
Attempt #1 failed!
False is not true

Attempt #2 failed!
False is not true

Attempt #3 failed!
False is not true

Attempt #4 failed!
False is not true

Attempt #5 failed!
False is not true

Attempt #6 failed!
False is not true

Attempt #7 failed!
False is not true

Attempt #8 failed!
False is not true

Attempt #9 failed!
False is not true

Attempt #10 failed!
False is not true
]]></system-out>
	</testcase>
	<testcase classname="pyspark.tests.test_util.UtilTests" name="test_find_spark_home" time="0.595" timestamp="2023-09-13T11:17:08" file="python/pyspark/tests/test_util.py" line="78"/>
	<testcase classname="pyspark.tests.test_util.UtilTests" name="test_parsing_version_string" time="0.617" timestamp="2023-09-13T11:17:08" file="python/pyspark/tests/test_util.py" line="73"/>
	<testcase classname="pyspark.tests.test_util.UtilTests" name="test_py4j_str" time="0.584" timestamp="2023-09-13T11:17:09" file="python/pyspark/tests/test_util.py" line="66"/>
</testsuite>
```

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/testing/utils.py [hunk 2]
```diff
@@ -80,7 +83,7 @@ def eventually(condition, timeout=30.0, catch_assertions=False):
     ----------
     condition : function
         Function that checks for termination conditions. condition() can return:
-            - True: Conditions met. Return without error.
+            - True or None: Conditions met. Return without error.
```
Comment (zhengruifeng) [inline python/pyspark/testing/utils.py:86]: @HyukjinKwon I made a change here, so that we don't need `return True` for most cases in the future.

**Comment by zhengruifeng** [comment]
merged to master

**Pull request closed without merging**